File: AIReaderKindle/src/Services/DictionaryPackWriter.cpp
#include "DictionaryPackWriter.hpp"

#include "../Domain/Dictionary/WordNormalizer.hpp"
#include "../Support/Files.hpp"
#include "Support/Json.hpp"
#include "../Support/Text.hpp"

#include <zlib.h>

DictionaryPackWriter::DictionaryPackWriter(const std::string& path) {
    Files::remove(path);
    database_ = std::make_unique<Database>(path);
    if (!database_->isOpen()) return;
    bool ok = database_->exec("CREATE TABLE metadata (key TEXT PRIMARY KEY, value TEXT NOT NULL)")
        && database_->exec("CREATE TABLE lemmas (id INTEGER PRIMARY KEY, word TEXT NOT NULL UNIQUE)")
        && database_->exec(
            "CREATE TABLE forms ("
            " normalized_form TEXT NOT NULL, ordinal INTEGER NOT NULL, form TEXT NOT NULL,"
            " lemma_id INTEGER NOT NULL, part_of_speech TEXT NOT NULL, gender TEXT, number TEXT,"
            " verb_info TEXT NOT NULL, PRIMARY KEY (normalized_form, ordinal)) WITHOUT ROWID")
        && database_->exec(
            "CREATE TABLE entries ("
            " lemma_id INTEGER NOT NULL, ordinal INTEGER NOT NULL, payload BLOB NOT NULL,"
            " PRIMARY KEY (lemma_id, ordinal)) WITHOUT ROWID")
        && database_->exec("BEGIN");
    if (!ok) database_.reset();
}

static std::string deflate(const std::string& json) {
    uLongf size = compressBound(static_cast<uLong>(json.size()));
    std::string out(size, '\0');
    if (compress(reinterpret_cast<Bytef*>(&out[0]), &size, reinterpret_cast<const Bytef*>(json.data()),
                 static_cast<uLong>(json.size())) != Z_OK) {
        return json;
    }
    out.resize(size);
    return out;
}

void DictionaryPackWriter::add(const DictionaryImportEntry& entry) {
    if (!isOpen()) return;
    std::string word = WordNormalizer::normalize(entry.headword);
    Json glosses = Json::array();
    for (const auto& sense : entry.senses) {
        std::string trimmed = Text::trim(sense);
        if (!trimmed.empty()) glosses.push(trimmed);
    }
    if (word.empty() || glosses.size() == 0) return;

    long long id;
    auto known = lemmaIds_.find(word);
    if (known != lemmaIds_.end()) {
        id = known->second;
    } else {
        id = static_cast<long long>(lemmaIds_.size()) + 1;
        lemmaIds_[word] = id;
        Statement insert(*database_, "INSERT INTO lemmas (id, word) VALUES (?, ?)");
        insert.bind(1, id).bind(2, word).run();
    }
    int ordinal = ordinals_[id]++;

    // The article shape `DictionaryDatabase` decodes.
    Json definition = Json::object();
    definition.set("glosses", glosses);
    Json article = Json::object();
    article.set("definitions", Json(std::vector<Json>{definition}));
    if (!entry.partOfSpeech.empty()) article.set("part_of_speech", entry.partOfSpeech);

    Statement insert(*database_, "INSERT INTO entries (lemma_id, ordinal, payload) VALUES (?, ?, ?)");
    insert.bind(1, id).bind(2, ordinal).bindBlob(3, deflate(article.dump())).run();
    ++entryCount_;
}

bool DictionaryPackWriter::finish(const std::string& targetLanguage, const std::string& definitionLanguage) {
    if (!isOpen() || entryCount_ == 0) return false;
    std::map<std::string, std::string> metadata = {
        {"schema_version", "2"},
        {"mode", "converted"},
        {"payload_encoding", "json+zlib"},
        {"lemma_count", std::to_string(lemmaIds_.size())},
        {"form_count", "0"},
        {"definition_entry_count", std::to_string(entryCount_)},
        {"target_language", targetLanguage},
        {"definition_language", definitionLanguage},
    };
    Statement insert(*database_, "INSERT INTO metadata (key, value) VALUES (?, ?)");
    for (const auto& pair : metadata) {
        if (pair.second.empty()) continue;
        insert.reset();
        insert.bind(1, pair.first).bind(2, pair.second).run();
    }
    return database_->exec("COMMIT");
}

File: AIReader/AIReader/Services/Dictionary/DictionaryPackWriter.swift
import Foundation
import SQLiteData

/// Writes the schema version 2 pack the app reads, one article at a time.
///
/// Converted dictionaries carry no inflection tables, so `forms` is left empty
/// and lookups fall back to matching the headword. That is why headwords are
/// stored normalized: the reader taps “Paris” and the query arrives as “paris”.
final class DictionaryPackWriter {
    enum WriteError: LocalizedError {
        case empty

        var errorDescription: String? {
            switch self {
            case .empty: "The file held no entries this app could read."
            }
        }
    }

    private let queue: DatabaseQueue
    private var lemmaIDs: [String: Int] = [:]
    private var ordinals: [Int: Int] = [:]
    private var pending: [(lemma: Int, ordinal: Int, payload: Data)] = []
    private var newLemmas: [(id: Int, word: String)] = []

    private(set) var entryCount = 0

    init(creating url: URL) throws {
        try? FileManager.default.removeItem(at: url)
        queue = try DatabaseQueue(path: url.path)
        try queue.write { db in
            try db.execute(sql: "CREATE TABLE metadata (key TEXT PRIMARY KEY, value TEXT NOT NULL)")
            try db.execute(sql: "CREATE TABLE lemmas (id INTEGER PRIMARY KEY, word TEXT NOT NULL UNIQUE)")
            try db.execute(sql: """
                CREATE TABLE forms (
                    normalized_form TEXT NOT NULL,
                    ordinal INTEGER NOT NULL,
                    form TEXT NOT NULL,
                    lemma_id INTEGER NOT NULL,
                    part_of_speech TEXT NOT NULL,
                    gender TEXT,
                    number TEXT,
                    verb_info TEXT NOT NULL,
                    PRIMARY KEY (normalized_form, ordinal)
                ) WITHOUT ROWID
                """)
            try db.execute(sql: """
                CREATE TABLE entries (
                    lemma_id INTEGER NOT NULL,
                    ordinal INTEGER NOT NULL,
                    payload BLOB NOT NULL,
                    PRIMARY KEY (lemma_id, ordinal)
                ) WITHOUT ROWID
                """)
        }
    }

    func add(_ entry: DictionaryImportEntry) throws {
        let word = WordNormalizer.normalize(entry.headword)
        let senses = entry.senses.map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
        guard !word.isEmpty, !senses.isEmpty else { return }

        let id = lemmaID(for: word)
        let ordinal = ordinals[id, default: 0]
        ordinals[id] = ordinal + 1

        guard let payload = payload(partOfSpeech: entry.partOfSpeech, senses: senses) else { return }
        pending.append((id, ordinal, payload))
        entryCount += 1
        if pending.count >= 2_000 { try flush() }
    }

    func finish(_ info: DictionaryImportInfo) throws {
        try flush()
        guard entryCount > 0 else { throw WriteError.empty }

        let metadata: [String: String] = [
            "schema_version": "2",
            "mode": "converted",
            "payload_encoding": "json+zlib",
            "lemma_count": String(lemmaIDs.count),
            "form_count": "0",
            "definition_entry_count": String(entryCount),
            "target_language": info.targetLanguage ?? "",
            "definition_language": info.definitionLanguage ?? ""
        ]
        try queue.write { db in
            for (key, value) in metadata where !value.isEmpty {
                try db.execute(sql: "INSERT INTO metadata (key, value) VALUES (?, ?)",
                               arguments: [key, value])
            }
        }
    }

    private func lemmaID(for word: String) -> Int {
        if let id = lemmaIDs[word] { return id }
        let id = lemmaIDs.count + 1
        lemmaIDs[word] = id
        newLemmas.append((id, word))
        return id
    }

    /// The article shape `DictionaryArticlePayload` decodes.
    private func payload(partOfSpeech: String?, senses: [String]) -> Data? {
        var article: [String: Any] = ["definitions": [["glosses": senses]]]
        if let partOfSpeech, !partOfSpeech.isEmpty { article["part_of_speech"] = partOfSpeech }
        guard let json = try? JSONSerialization.data(withJSONObject: article) else { return nil }
        return Deflate.zlib(json) ?? json
    }

    private func flush() throws {
        guard !pending.isEmpty || !newLemmas.isEmpty else { return }
        let lemmas = newLemmas
        let entries = pending
        newLemmas = []
        pending = []

        try queue.write { db in
            for lemma in lemmas {
                try db.execute(sql: "INSERT INTO lemmas (id, word) VALUES (?, ?)",
                               arguments: [lemma.id, lemma.word])
            }
            for entry in entries {
                try db.execute(
                    sql: "INSERT INTO entries (lemma_id, ordinal, payload) VALUES (?, ?, ?)",
                    arguments: [entry.lemma, entry.ordinal, entry.payload]
                )
            }
        }
    }
}
